oauthenticator/oauth2.py

```diff
@@ -1025,6 +1047,18 @@ async def check_allowed(self, username, auth_model):
         if username in self.allowed_users:
             return True
 
+        # If we specific scope grants are required, validate that they have been granted
```

Comment:
I've considered if this should be in an earlier stage than check_allowed. Its nice to have it here if seen as authorization logic, but it could be seen as a technical requirement - for example to get relevant info for the username. Then, it would be more suitable that not meeting this criteria is seen ss an error rather than 403.

Hmmmm... Consider if we request email scope and username_claim email and declare it as email scope required but don't get it - should that be another kind of error rather than denied authorization 403?

If we think so, this could be more like an assertion early in update_auth_model instead raising an error maybe? Wdyt?

Reply:
I'm not sure what i think, but it wouldn't be breaking to relocate this check so I'm open with going onwards either way

Reply:
Good question. There are two ways to think about this:

1. It's part of *authentication* logic, as if we aren't granted a specific *claim* from a scope, we can't even log the user in (for username_claim, for example). So if the scope isn't granted, we can't actually validate who this user is. So it gets relocated elsewhere.
2. It is part of *authorization* logic, where scopes are used as a way to 'gate' access *after* we could validate who the user is. So we *know* who the user is, but we don't want to give them access. In this case it stays where it is.

So how do we best determine if this is part of authentication or authorization?

I went to look at possible available scopes for a bunch of authenticators:

- [GitHub scopes list](https://docs.github.com/en/apps/oauth-apps/building-oauth-apps/scopes-for-oauth-apps)
- [Mediawiki rights](https://www.mediawiki.org/wiki/Extension:OAuth/en#User_rights)
- [Standard OpenID connect scopes](https://auth0.com/docs/get-started/apis/scopes/openid-connect-scopes#standard-claims) (as supported by Auth0)

`username_claim` feels clearly part of authentication to me, and right now, if we don't get a `username_claim` back for any reason, it's already a 403. But the 403 is a little bit unclear on why exactly the permission was denied and can be frustrating for the user. However, the scopes that can be related to `username_claim` are few in number. In GitHub for example, it really is just `user` - its unlikely you would want the username to be derived from anything else (for the most part). I think this is generally true for most auth providers, particularly given the prevalance & design decisions of openid connect. This means that when configuring, if the `username_claim` is not present due to a missing `scope`, it *most likely* is because the hub admin did not *ask* for the scope, and it's going to be clear to them in initial testing.

However, you may want to deny *authorization* for some subsets - for example, someone may want to setup a specific ssh key in github for a JupyterHub, and hence reject authorization if they don't grant `write:public_key`, even though it has nothing to do with authentication itself - we know who the user is, just don't want them in here.

I'm a big fan of @GeorgianaElena's work in https://github.com/jupyterhub/oauthenticator/pull/594, and think that separation of authentication and authorization here makes things easier to reason about. I think the most likely use of this feature is *authorization* and not *authentication*. `username_claim` being set to something that isn't present if a specific scope is not granted is IMO a different problem, and we can tackle that differently if you desire - I'd say the primary extra bit there is to find a way to provide useful error information. 

All this to say, I think it's ok for this check to remain here :)

Reply:
I opened https://github.com/jupyterhub/oauthenticator/pull/721 to make the missing username_claim situation slightly easier.

Reply:
Thanks for reasoning about this, I think this makes a good case for considering this to be part of authorization (authz) logic rather than authentication (authn) logic!